File: solutions/typescript/0008.ts
function myAtoi(s: string): number {
  const MAX_INT = Math.pow(2, 31) - 1;
  const MIN_INT = -Math.pow(2, 31);

  // 去除前导空格
  s = s.trim();
  if (s.length === 0) return 0;

  let sign = 1;
  let i = 0;

  // 处理正负号
  if (s[i] === "+" || s[i] === "-") {
    sign = s[i] === "-" ? -1 : 1;
    i++;
  }

  let result = 0;

  // 处理数字
  while (i < s.length && /^\d$/.test(s[i])) {
    const digit = parseInt(s[i]);

    // 检查是否溢出
    if (
      result > Math.floor(MAX_INT / 10) ||
      (result === Math.floor(MAX_INT / 10) && digit > MAX_INT % 10)
    ) {
      return sign === 1 ? MAX_INT : MIN_INT;
    }

    result = result * 10 + digit;
    i++;
  }

  return result * sign;
}

File: solutions/go/0008.go
func myAtoi(s string) int {
	MAX_INT := 1<<31 - 1
	MIN_INT := -1 << 31
	s = strings.TrimSpace(s)
	if s == "" {
		return 0
	}
	sign := 1
	if s[0] == '-' {
		sign = -1
		s = s[1:]
	} else if s[0] == '+' {
		s = s[1:]
	}
	result := 0
	for _, char := range s {
		if !unicode.IsDigit(char) {
			break
		}
		digit := int(char - '0')
		if sign == 1 && (result > MAX_INT/10 || (result == MAX_INT/10 && digit > MAX_INT%10)) {
			return MAX_INT
		}
		if sign == -1 && (result > -(MIN_INT/10) || (result == -(MIN_INT/10) && digit > -(MIN_INT%10))) {
			return MIN_INT
		}
		result = result*10 + digit
	}
	return result * sign
}